drools-ansible-rulebook-integration-runtime/src/main/java/org/drools/ansible/rulebook/integration/core/jpy/AsyncAstRulesEngine.java

```diff
@@ -0,0 +1,128 @@
+package org.drools.ansible.rulebook.integration.core.jpy;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.drools.ansible.rulebook.integration.api.RuleFormat;
+import org.drools.ansible.rulebook.integration.api.RuleNotation;
+import org.drools.ansible.rulebook.integration.api.domain.RulesSet;
+import org.json.JSONObject;
+
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.StandardSocketOptions;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import static org.drools.ansible.rulebook.integration.api.RulesExecutor.OBJECT_MAPPER;
+
+public class AsyncAstRulesEngine {
+    static class Response {
+        private final long session_id;
+        private final Object result;
+
+        public Response(long session_id, Object result) {
+            this.session_id = session_id;
+            this.result = result;
+        }
+
+        public long getSession_id() {
+            return session_id;
+        }
+
+        public Object getResult() {
+            return result;
+        }
+    }
+
+    private final ServerSocket socketChannel;
+
+    private final AstRulesEngineInternal astRulesEngine;
+    private DataOutputStream ssc;
+
+    public AsyncAstRulesEngine() {
+        try {
+            astRulesEngine = new AstRulesEngineInternal();
+            socketChannel = new ServerSocket(0);
+            socketChannel.setOption(StandardSocketOptions.SO_REUSEADDR, true);
+            accept();
+            System.out.printf("Server started at %s.\n", port());
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public int port() {
+        return socketChannel.getLocalPort();
+    }
+
+    private void accept() {
+        CompletableFuture.supplyAsync(() -> {
+            try {
+                return socketChannel.accept();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }).handleAsync((skt, thrw) -> {
+            try {
+                this.ssc = new DataOutputStream(skt.getOutputStream());
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            return null;
+        });
+
+    }
+
+    public void createRuleset(String rulesetString) {
+        RulesSet rulesSet = RuleNotation.CoreNotation.INSTANCE.toRulesSet(RuleFormat.JSON, rulesetString);
+        long ruleset = astRulesEngine.createRuleset(rulesSet);
+        try {
+            ssc.writeInt(8);
```

Comment:
@evacchi Why do we need the session id to be coming on the socket. We would need that call to be synchronous. There is no waiting when we createRuleset.

Reply:
sure, I can keep that synchronous